rfcs/0001-inline-expression-value-logging.md

```diff
@@ -0,0 +1,191 @@
+  without needing to restructure using additional `let` bindings, or
+  by duplicating sub-expressions.
+
+* For debug/trace level logging in particular, automatic formatting of
+  the expression with its value can give sufficient context in the log
+  (particularly with _target_, _file_ and _line_ information) and avoids
+  additional boilerplate.
+
+In a project where configuring `log` and an output logger
+implementation hasn't (yet) happened, one can conveniently use
+[`std::dbg!`](https://doc.rust-lang.org/std/macro.dbg.html), with no
+initial setup cost, for `expression = value` printing to STDERR. From
+the linked rustdoc:
+
+> Note that the macro is intended as a debugging tool and therefore
+> you should avoid having uses of it in version control for longer
+> periods. Use cases involving debug output that should be added to
+> version control may be better served by macros such as `debug!` from
+> the `log` crate.
+
+Indeed, a major point of the `log` package and `Level`s, is the
+ability to keep `debug!` and `trace!` logging in place for further
+use, including by other contributors, without paying a cost for
+unlogged messages in release builds.
+
+It follows that for projects that _do already have_ `log` and an
+output logger dependency and configuration, and particularly with some
+debug/trace logging already in place, use of `std::dbg!` would be
+unwelcome in PRs, and likely less productive than a `log`-based
+equivalent. For parity of convenience, the `log` API needs an
+extension for logging expressions and passing through values inline.
+
+# Detailed Design
+
+In addition to the existing set of _formatted logging_ macros, by
+level, e.g. `trace!`, `debug!`, etc., we add a new set of _inline
+expression and value logging_ macros, with a “_-v_” suffix: `tracev!`,
+`debugv!`, and so forth for all levels. The _-v_ macros take a _single_
+expression argument, which is evaluated exactly once, regardless of if
+the logging level is enabled or not, and returned:
+
+```rust
+let n = 12;
+let m = debugv!(n / 2) - 1;
+//      ^-- debug log message: "n / 2 = 6"
+assert_eq!(m, 5);
+```
+
+The _default_ format string for the _-v_ macros is `"{} = {:?}"`,
+where the `stringify!`-ed expression and resulting value are passed,
+in that order.  Note that the `std::dbg!` macro currently uses `"{} =
+{:#?}"`—the value is "pretty-printed", potentially over multiple
+lines.  Given the line-orientation of logging output, the default
+format for the _-v_ macros avoids this. However, a custom format may
+also be passed before the expression, which adds considerably more
+output flexibility:
+
+```rust
+let i = 32;
+tracev!("{} = {}", i);            // use `Display` instead of `Debug`
+tracev!("{} = {:x}", i);          // hexadecimal format value
+tracev!("{} = {:#?}", i);         // use pretty, multi-line format
+
+let rem = infov!("{1:?} remaining ({0})", deadline - Instance::now());
```

Comment:
I quite like how this has turned out :+1:

Reply:
Thinking about real-world examples more, I'm guessing these macros would be useful as a way to 'annotate' an expression to trace your way through interesting operations. That suggests we might actually want the expression and format to be reversed in the macro so the most relevant info when scanning the code appears first:

```rust
let rem = deadline - Instance::now();
```

becomes:

```rust
let rem = tracev!(deadline = Instance::now(), "{value:?} remaining ({expr})");
```

where we could interpolate the expression and value as named args into the format.

Reply:
I actually had them named in my implementation branch at one point. It's easy to do. As you likely know, you can still use the position references,  `"{1} {0}"` even with named arguments, so I could resurect that in the implementation or you could add it later.

Reply:
Also, in my own limited experimentation I like using "{0} -> {:?}" as the format string and I wonder if that would be a better default format, as "->" doesn't appear in any expressions I'm aware of, but can of course be used for assignment in an expression position. Thoughts?

Reply:
Naming format args is no longer relevant (current RFC) given that only the value format is customizable.   Parameter order under consideration below.